### PR TITLE
TSV-826: Amended to accept either a nino or a postcode.

### PR DESCRIPTION
**Pull request by steve-edmans**

### CODE REVIEW

File: app/uk/gov/hmrc/personaldetailsvalidation/PersonalDetailsValidationResourceController.scala [hunk 1]
```diff
@@ -47,10 +47,19 @@ class PersonalDetailsValidationResourceController @Inject()(personalDetailsValid
   def create = Action.async(parse.json) { implicit request =>
     withJsonBody[PersonalDetails] { personalDetails =>
 
-      def handleMatchingDone(validationId: ValidationId) =
+      def handleMatchingDone(validationId: ValidationId) = {
         Future.successful(Created.withHeaders(LOCATION -> routes.PersonalDetailsValidationResourceController.get(validationId).url))
+      }
 
-      def handleException(exception: Exception): Future[Result] = Future.failed(exception)
+      def handleException(exception: Exception): Future[Result] = {
```
Comment (anshulbajpai) [inline app/uk/gov/hmrc/personaldetailsvalidation/PersonalDetailsValidationResourceController.scala]: This code can be avoided if validations are done as part of Reads.

File: app/uk/gov/hmrc/personaldetailsvalidation/PersonalDetailsValidator.scala [hunk 2]
```diff
@@ -52,16 +52,28 @@ private class PersonalDetailsValidator[Interpretation[_] : Monad](matchingConnec
...
       personalDetailsValidation = matchResult.toPersonalDetailsValidation(optionallyHaving = personalDetails)
       _ <- personalDetailsValidationRepository.create(personalDetailsValidation)
       _ = sendEvents(matchResult, personalDetails)
     } yield personalDetailsValidation.id
-  }.leftMap { error => sendErrorEvents(personalDetails); error }
+  }.leftMap {
+    error => sendErrorEvents(personalDetails); error
+  }
+
+  private def validateContentRules(personalDetails: PersonalDetails): EitherT[Interpretation, Exception, PersonalDetails] = {
```
Comment (anshulbajpai) [inline app/uk/gov/hmrc/personaldetailsvalidation/PersonalDetailsValidator.scala]: This validation can move into Reads

File: app/uk/gov/hmrc/personaldetailsvalidation/PersonalDetailsValidator.scala [hunk 3]
```diff
@@ -52,16 +52,28 @@ private class PersonalDetailsValidator[Interpretation[_] : Monad](matchingConnec
...
+    personalDetails match {
+      case bad if (bad.nino.isDefined && bad.postCode.isDefined) => EitherT.leftT[Interpretation, PersonalDetails](new IllegalArgumentException("both nino and postcode supplied"))
```
Comment (anshulbajpai) [inline app/uk/gov/hmrc/personaldetailsvalidation/PersonalDetailsValidator.scala]: This validation can move into the Reads

File: app/uk/gov/hmrc/personaldetailsvalidation/audit/AuditDataEventFactory.scala [hunk 4]
```diff
@@ -43,12 +43,24 @@ private[personaldetailsvalidation] class AuditDataEventFactory(auditConfig: Audi
...
+    val postCode = personalDetails.postCode match {
+      case Some(postCodeValue) => postCodeValue
+      case None => "NOT SUPPLIED"
+    }
+
+    DataEvent(
+      auditSource = auditConfig.appName,
+      auditType = auditType,
+      tags = auditTagProvider(hc, auditType, request),
+      detail = auditDetailsProvider(hc) + ("nino" -> nino) + ("postCode" -> postCode) + ("matchingStatus" -> matchingStatus)
```
Comment (anshulbajpai) [inline app/uk/gov/hmrc/personaldetailsvalidation/audit/AuditDataEventFactory.scala]: We don't need to audit both the nino and postcode for the same validation request. We know it is going to be either of the cases.

File: app/uk/gov/hmrc/personaldetailsvalidation/model/PersonalDetails.scala [hunk 5]
```diff
@@ -24,4 +24,5 @@ import uk.gov.hmrc.model.NonEmptyString
 case class PersonalDetails(firstName: NonEmptyString,
```
Comment (anshulbajpai) [inline app/uk/gov/hmrc/personaldetailsvalidation/model/PersonalDetails.scala]: This domain object is very loose now. It is easy to create an instance of `PersonalDetails ` case class by providing both nino and postcode.
I propose having an ADT (algebraic data type) structure with two case classes `PersonalDetailsWithPostcode` and `PersonalDetailsWithNino` (or better names which can convey the right meaning) having a common base trait `PersonalDetails`. This will not let anybody create invalid domain objects.

File: app/uk/gov/hmrc/personaldetailsvalidation/PersonalDetailsValidationResourceController.scala [hunk 6]
```diff
@@ -70,7 +79,16 @@ object PersonalDetailsValidationResourceController {
     (__ \ "firstName").readOrError[String]("firstName is missing").filter(ValidationError("firstName is blank/empty"))(_.trim.nonEmpty) and
       (__ \ "lastName").readOrError[String]("lastName is missing").filter(ValidationError("lastName is blank/empty"))(_.trim.nonEmpty) and
       (__ \ "dateOfBirth").readOrError[LocalDate]("dateOfBirth is missing/invalid") and
-      (__ \ "nino").readOrError[String]("nino is missing").map(_.toUpperCase.replaceAll("""\s""", "")).filter(ValidationError("invalid nino format"))(str => Try(Nino(str)).isSuccess).map(Nino)
-    ) (PersonalDetails(_, _, _, _))
-
+      (__ \ "nino").readNullable[String].map {
```
Comment (anshulbajpai) [inline app/uk/gov/hmrc/personaldetailsvalidation/PersonalDetailsValidationResourceController.scala]: This Reads code can be modified to return either of the instances of the ADT structure I mentioned here - https://github.com/hmrc/personal-details-validation/pull/13/files#r171514238.

File: app/uk/gov/hmrc/personaldetailsvalidation/PersonalDetailsValidationResourceController.scala (hunk 1 above)
Comment (jachro) [reply app/uk/gov/hmrc/personaldetailsvalidation/PersonalDetailsValidationResourceController.scala]: 👍

File: app/uk/gov/hmrc/personaldetailsvalidation/PersonalDetailsValidator.scala (hunk 3 above)
Comment (jachro) [reply app/uk/gov/hmrc/personaldetailsvalidation/PersonalDetailsValidator.scala]: I agree with Anshul. This will take lines of code from here and allow to use Play mechanisms to turn form validation errors to BAD_REQUEST. Actually, the logic of extracting domain objects from request's payload belongs to the form binding and it seems reasonable that it does not leak out of the page PersonalDetailsPage.

File: app/uk/gov/hmrc/personaldetailsvalidation/audit/AuditDataEventFactory.scala (hunk 4 above)
Comment (jachro) [reply app/uk/gov/hmrc/personaldetailsvalidation/audit/AuditDataEventFactory.scala]: I guess once we go for the ADT solution Anshul mentioned below, we could have two different strategies implicitly injected into the 'createEvent' method to cope with both of the types.

File: app/uk/gov/hmrc/personaldetailsvalidation/model/PersonalDetails.scala (hunk 5 above)
Comment (jachro) [reply app/uk/gov/hmrc/personaldetailsvalidation/model/PersonalDetails.scala]: Few more thing for using separate types for both cases would be that we will have less 'if' conditions spread in different places, more compile-time safety and then fewer unit tests. Effectively, because the requirement lists two different excluding cases which have to be dealt with in slightly different ways, it somehow seems to me more natural to model it with two types.

File: app/uk/gov/hmrc/personaldetailsvalidation/matching/MatchingConnector.scala [hunk 7]
```diff
@@ -64,12 +64,10 @@ class FuturedMatchingConnector @Inject()(httpClient: HttpClient, connectorConfig
   }
 
   private implicit class PersonalDetailsSerializer(personalDetails: PersonalDetails) {
-    lazy val toJson: JsObject = Json.obj(
```
Comment (anshulbajpai) [inline app/uk/gov/hmrc/personaldetailsvalidation/matching/MatchingConnector.scala]: We choose not to use `Json.format` because we want to control what data we are sending out to the authenticator. This future-proofs us by not sending additional fields (which we can add in future) in PersonalDetails case class.

File: app/uk/gov/hmrc/personaldetailsvalidation/matching/MatchingConnector.scala [hunk 8]
```diff
@@ -64,12 +64,10 @@ class FuturedMatchingConnector @Inject()(httpClient: HttpClient, connectorConfig
...
-      "firstName" -> personalDetails.firstName,
-      "lastName" -> personalDetails.lastName,
-      "dateOfBirth" -> personalDetails.dateOfBirth,
-      "nino" -> personalDetails.nino
-    )
+    lazy val toJson: JsObject = {
```
Comment (anshulbajpai) [inline app/uk/gov/hmrc/personaldetailsvalidation/matching/MatchingConnector.scala]: We choose not to use Json.format because we want to control what data we are sending out to the authenticator. This future-proofs us by not sending additional fields (which we can add in future) in PersonalDetails case class.

File: app/uk/gov/hmrc/personaldetailsvalidation/model/PersonalDetails.scala (hunk 5 above)
Comment (steve-edmans) [reply app/uk/gov/hmrc/personaldetailsvalidation/model/PersonalDetails.scala]: We have the requirement to update the validation resource with the Nino, so there will be cases where we have nino and postcode.

Comment (anshulbajpai) [reply app/uk/gov/hmrc/personaldetailsvalidation/model/PersonalDetails.scala]: @steve-edmans - The Validation resource itself will contain only the NINO. The input request body will either contain the postcode or the NINO. I don't see how we can have both of them together.

File: app/uk/gov/hmrc/personaldetailsvalidation/formats/PersonalDetailsExternalFormat.scala [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+        }.filter(ValidationError("invalid nino format")){
+          case Some(nino) => Try(Nino(nino)).isSuccess
+          case _ => true
+        }.map{
+          case Some(nino) => Some(Nino(nino))
+          case _ => None
+        } and
+          (__ \ "postCode").readNullable[String]
+        ).tupled.
+        filter(ValidationError("at least nino or postcode needs to be supplioed supplied")){
```
Comment (anshulbajpai) [inline app/uk/gov/hmrc/personaldetailsvalidation/formats/PersonalDetailsExternalFormat.scala]: the message has `supplied` word twice and first occurrence has a typo.

File: app/uk/gov/hmrc/personaldetailsvalidation/formats/PersonalDetailsExternalFormat.scala [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+        }.
+        filter(ValidationError("both nino and postcode supplied")){
+          case (Some(_), Some(_)) => false
+          case _ => true
+        }
+    )((firstName, lastName, dateOfBirth, ninoOrPostCode) => {
+    ninoOrPostCode match {
+      case (Some(nino), None) => PersonalDetailsWithNino(firstName, lastName, dateOfBirth, nino)
+      case (None, Some(postCode)) => PersonalDetailsWithPostCode(firstName, lastName, dateOfBirth, postCode)
+      case (Some(nino), Some(postCode)) => PersonalDetailsWithNinoAndPostCode(firstName, lastName, dateOfBirth, nino, postCode)
```
Comment (anshulbajpai) [inline app/uk/gov/hmrc/personaldetailsvalidation/formats/PersonalDetailsExternalFormat.scala]: I think this case is not possible once we do the validation.

File: app/uk/gov/hmrc/personaldetailsvalidation/PersonalDetailsValidationResourceController.scala [hunk 11]
```diff
@@ -16,41 +16,44 @@
...
   def create = Action.async(parse.json) { implicit request =>
     withJsonBody[PersonalDetails] { personalDetails =>
 
-      def handleMatchingDone(validationId: ValidationId) =
+      def handleMatchingDone(validationId: ValidationId) = {
         Future.successful(Created.withHeaders(LOCATION -> routes.PersonalDetailsValidationResourceController.get(validationId).url))
+      }
 
-      def handleException(exception: Exception): Future[Result] = Future.failed(exception)
+      def handleException(exception: Exception): Future[Result] = {
```
Comment (anshulbajpai) [inline app/uk/gov/hmrc/personaldetailsvalidation/PersonalDetailsValidationResourceController.scala]: Is this change still required?

File: app/uk/gov/hmrc/personaldetailsvalidation/model/PersonalDetails.scala [hunk 12]
```diff
@@ -18,10 +18,68 @@ package uk.gov.hmrc.personaldetailsvalidation.model
...
+
+  lazy val toJson: JsObject = Json.obj(
+    "firstName" -> firstName,
+    "lastName" -> lastName,
+    "dateOfBirth" -> dateOfBirth,
+    "postCode" -> postCode.value
+  )
+}
+
+case class PersonalDetailsWithNinoAndPostCode(firstName: NonEmptyString,
```
Comment (anshulbajpai) [inline app/uk/gov/hmrc/personaldetailsvalidation/model/PersonalDetails.scala]: I don't think there is a requirement to expose postcode on the GET endpoint which makes the need of persisting postcode redundant.

File: app/uk/gov/hmrc/personaldetailsvalidation/PersonalDetailsValidator.scala [hunk 13]
```diff
@@ -60,9 +60,13 @@ private class PersonalDetailsValidator[Interpretation[_] : Monad](matchingConnec
   }.leftMap { error => sendErrorEvents(personalDetails); error }
 
   private implicit class MatchResultOps(matchResult: MatchResult) {
-    def toPersonalDetailsValidation(optionallyHaving: PersonalDetails): PersonalDetailsValidation = matchResult match {
-      case MatchSuccessful(_) => PersonalDetailsValidation.successful(optionallyHaving)
-      case MatchFailed(_) => PersonalDetailsValidation.failed()
+    def toPersonalDetailsValidation(optionallyHaving: PersonalDetails): PersonalDetailsValidation = {
+      (matchResult, optionallyHaving) match {
+        case (MatchSuccessful(matchingPerson: PersonalDetailsNino), other: PersonalDetailsWithPostCode) =>
+          PersonalDetailsValidation.successful(other.addNino(matchingPerson.nino))
```
Comment (anshulbajpai) [inline app/uk/gov/hmrc/personaldetailsvalidation/PersonalDetailsValidator.scala]: PersonalDetailsValidation object graph doesn't need postcode. Refer this - https://github.com/hmrc/personal-details-validation/pull/13/files#r172578403

File: app/uk/gov/hmrc/personaldetailsvalidation/formats/PersonalDetailsInternalFormat.scala [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+          case Some(nino) => Some(Nino(nino))
+          case _ => None
+        } and
+          (__ \ "postCode").readNullable[String]
+        ).tupled
+    )((firstName, lastName, dateOfBirth, ninoOrPostCode) => {
+    ninoOrPostCode match {
+      case (Some(nino), None) => PersonalDetailsWithNino(firstName, lastName, dateOfBirth, nino)
+      case (None, Some(postCode)) => PersonalDetailsWithPostCode(firstName, lastName, dateOfBirth, postCode)
+      case (Some(nino), Some(postCode)) => PersonalDetailsWithNinoAndPostCode(firstName, lastName, dateOfBirth, nino, postCode)
```
Comment (anshulbajpai) [inline app/uk/gov/hmrc/personaldetailsvalidation/formats/PersonalDetailsInternalFormat.scala]: We shouldn't be persisting postcode in first place.

File: app/uk/gov/hmrc/personaldetailsvalidation/model/PersonalDetails.scala [hunk 15]
```diff
@@ -18,10 +18,68 @@ package uk.gov.hmrc.personaldetailsvalidation.model
 
 import java.time.LocalDate
 
+import play.api.libs.json._
 import uk.gov.hmrc.domain.Nino
 import uk.gov.hmrc.model.NonEmptyString
 
-case class PersonalDetails(firstName: NonEmptyString,
+trait PersonalDetails {
+  def toJson: JsObject
```
Comment (anshulbajpai) [inline app/uk/gov/hmrc/personaldetailsvalidation/model/PersonalDetails.scala:26]: I don't feel we should be having `toJson` on domain objects. 
You can possibly create extension methods using implicits.

File: it/uk/gov/hmrc/personaldetailsvalidation/PersonalDetailsValidationISpec.scala [hunk 16]
```diff
@@ -106,6 +142,36 @@ class PersonalDetailsValidationISpec extends BaseIntegrationSpec {
...
+      """
+        |{
+        |   "firstName": "Jim",
+        |   "lastName": "Ferguson",
+        |   "postCode": "SE1 9NT",
+        |   "dateOfBirth": "1948-04-23"
+        |}
+      """.stripMargin
+
+    val personalDetailsBoth =
```
Comment (anshulbajpai) [inline it/uk/gov/hmrc/personaldetailsvalidation/PersonalDetailsValidationISpec.scala]: Can we rename it to personalDetailsWithBothNinoAndPostcode

File: it/uk/gov/hmrc/personaldetailsvalidation/PersonalDetailsValidationISpec.scala [hunk 17]
```diff
@@ -106,6 +142,36 @@ class PersonalDetailsValidationISpec extends BaseIntegrationSpec {
...
+        |{
+        |   "firstName": "Jim",
+        |   "lastName": "Ferguson",
+        |   "nino": "AA000003D",
+        |   "postCode": "SE1 9NT",
+        |   "dateOfBirth": "1948-04-23"
+        |}
+      """.stripMargin
+
+    val invalidPersonalDetailsNeither =
```
Comment (anshulbajpai) [inline it/uk/gov/hmrc/personaldetailsvalidation/PersonalDetailsValidationISpec.scala]: Can we rename it to invalidPersonalDetailsWithNeitherPostcodeOrNino?

File: test/uk/gov/hmrc/personaldetailsvalidation/PersonalDetailsValidationResourceControllerSpec.scala [hunk 18]
```diff
@@ -79,7 +79,7 @@ class PersonalDetailsValidationResourceControllerSpec
     forAll(ninoTransformationScenarios) { (scenario, originalNinoValue, finalNinoValue) =>
       s"$scenario" in new Setup {
 
-        val requestPersonalDetails = randomPersonalDetails
+        val requestPersonalDetails : PersonalDetailsWithNino = randomPersonalDetails.asInstanceOf[PersonalDetailsWithNino]
```
Comment (anshulbajpai) [inline test/uk/gov/hmrc/personaldetailsvalidation/PersonalDetailsValidationResourceControllerSpec.scala:82]: Why is typecast required?

File: test/uk/gov/hmrc/personaldetailsvalidation/audit/AuditDataEventFactorySpec.scala [hunk 19]
```diff
@@ -25,13 +25,14 @@ import play.api.test.FakeRequest
...
 import uk.gov.hmrc.personaldetailsvalidation.matching.MatchingConnector.MatchResult.{MatchFailed, MatchSuccessful}
+import uk.gov.hmrc.personaldetailsvalidation.model._
 import uk.gov.hmrc.play.test.UnitSpec
 
 class AuditDataEventFactorySpec extends UnitSpec with MockFactory {
 
   "factory" should {
 
-    val personalDetails = personalDetailsObjects.generateOne
+    val personalDetails : PersonalDetailsWithNino = personalDetailsObjects.generateOne.asInstanceOf[PersonalDetailsWithNino]
```
Comment (anshulbajpai) [inline test/uk/gov/hmrc/personaldetailsvalidation/audit/AuditDataEventFactorySpec.scala:35]: typecast?

File: app/uk/gov/hmrc/personaldetailsvalidation/model/PersonalDetails.scala (hunk 15 above)
Comment (steve-edmans) [reply app/uk/gov/hmrc/personaldetailsvalidation/model/PersonalDetails.scala:26]: I disagree, we should not be adding layers of abstraction just for the sake of adding layers of abstraction.

IMO you should not create extension methods, except for objects that you do not own.

File: test/uk/gov/hmrc/personaldetailsvalidation/PersonalDetailsValidationResourceControllerSpec.scala (hunk 18 above)
Comment (steve-edmans) [reply test/uk/gov/hmrc/personaldetailsvalidation/PersonalDetailsValidationResourceControllerSpec.scala:82]: They are there to remove the "public method that needs an explicit type" warning.

File: app/uk/gov/hmrc/personaldetailsvalidation/model/PersonalDetails.scala (hunk 15 above)
Comment (anshulbajpai) [reply app/uk/gov/hmrc/personaldetailsvalidation/model/PersonalDetails.scala:26]: One of the aspects of functional programming is to keep functions separated from data. We usually try to follow that whenever we can. It is a matter of preference.
Also, even from OOP aspect, just because a class has data, it doesn't mean we should be creating any method on that class. Usually, domain classes should have domain related methods.
I leave it upto you how you want to do it.

File: test/uk/gov/hmrc/personaldetailsvalidation/PersonalDetailsValidationResourceControllerSpec.scala (hunk 18 above)
Comment (anshulbajpai) [reply test/uk/gov/hmrc/personaldetailsvalidation/PersonalDetailsValidationResourceControllerSpec.scala:82]: I think we should then change the type of `randomPersonalDetails` from PersonalDetails to `PersonalDetailsWithNino` because that is what is getting created.

File: test/uk/gov/hmrc/personaldetailsvalidation/audit/AuditDataEventFactorySpec.scala [hunk 20]
```diff
@@ -45,7 +46,7 @@ class AuditDataEventFactorySpec extends UnitSpec with MockFactory {
         dataEvent.auditSource shouldBe auditConfig.appName
         dataEvent.auditType shouldBe "MatchingResult"
         dataEvent.tags shouldBe auditTags
-        dataEvent.detail shouldBe auditDetails + ("nino" -> personalDetails.nino.value) ++ matchingingDetails
+        dataEvent.detail shouldBe auditDetails + ("nino" -> personalDetails.nino.value) + ("postCode" -> "NOT SUPPLIED") ++ matchingingDetails
```
Comment (anshulbajpai) [inline test/uk/gov/hmrc/personaldetailsvalidation/audit/AuditDataEventFactorySpec.scala:55]: I think we need to add two tests for the cases when the postcode is provided 
1.  successful matching 
2. failed matching

File: test/uk/gov/hmrc/personaldetailsvalidation/PersonalDetailsValidationResourceControllerSpec.scala [hunk 21]
```diff
@@ -135,7 +135,7 @@ class PersonalDetailsValidationResourceControllerSpec
         "firstName is missing",
         "lastName is missing",
         "dateOfBirth is missing/invalid",
-        "nino is missing"
+        "at least nino or postcode needs to be supplied"
```
Comment (anshulbajpai) [inline test/uk/gov/hmrc/personaldetailsvalidation/PersonalDetailsValidationResourceControllerSpec.scala:138]: I couldn't find test cases for postcode and nino mutually exclusive validation. Are they in some other file?

Comment (anshulbajpai) [reply test/uk/gov/hmrc/personaldetailsvalidation/PersonalDetailsValidationResourceControllerSpec.scala:138]: I see you have written integration tests but I think unit tests are helpful when you need faster feedback from the tests.

File: app/uk/gov/hmrc/personaldetailsvalidation/audit/EventsSender.scala [hunk 22]
```diff
@@ -61,7 +61,12 @@ private[personaldetailsvalidation] class EventsSender @Inject()(platformAnalytic
   }
 
   private implicit class PersonalDetailsOps(target: PersonalDetails) {
-    def hasSameNinoSuffixAs(other: PersonalDetails): Boolean = target.nino.value.last == other.nino.value.last
+    def hasSameNinoSuffixAs(other: PersonalDetails): Boolean = {
+      (target, other) match {
+        case (first : PersonalDetailsNino, second : PersonalDetailsNino) => first.nino.value == second.nino.value
+        case _ => false
```
Comment (anshulbajpai) [inline app/uk/gov/hmrc/personaldetailsvalidation/audit/EventsSender.scala:67]: This case can end up sending GA event `success_nino_suffix_different_from_cid ` when postcode is provided as a user input and it is a successful matching.
We should not be sending GA event in this case.

Comment (steve-edmans) [reply app/uk/gov/hmrc/personaldetailsvalidation/audit/EventsSender.scala:67]: Isn't this part of TSV-952? I think we would need input from the BAs to determine the exact GA events that need to be created.
As it is we will log a "success_nino_suffix_different_from_cid" even if they pass in a postcode.
